# W2D2 Notes

## Requiring Files
To require additional files into your class, you can use **require_relative** followed by the file name you are requiring.

```ruby
#when requiring files in the same directory
require_relative "./cat.rb"
#when requiring files in a inner directory
require_relative "./other_animals/dog.rb"
```

## User Input
To get user input, you can use **gets**. Gets has a problem that it will return user input with a new line. TO remove the line, we can use the **#chomp** method. If you are accepting integers instead of a string, be sure to use the method **#to_i**.

```ruby
#when accepting a string
input = gets.chomp
#when accepting a integer
input = gets.to_i
```

## Abstraction and Encapsulation
**OOP** (Object Oriented Programming) is a design pattern developers use when building application. OOP goal is to create programs using objects that interact with each other. 2 pillars of OOP are **Abstraction** and **Encapsulation**.

**Abstraction** is exposing essential features of an object while hiding inner details that are not necessary to *using* the feature.

**Encapsulation** closely relates methods and data attributes together with the hope of preventing misuse. 

## Attr Methods
attr_reader is the same as a getter method.
attr_writer is the same as a setter method.
attr_accessor is both a getter and setter method.
They are all written before the initialize method and used to make our code look cleaner.

## Syntatic Sugar
**Syntatic Sugar** is known as a "shortcut" in code.
```ruby
#the second statement is syntatic sugar due to it being more elegant
p person_1.==(person_2)
p person_1 == person_2
#the second statement is syntatic sugar due to it being more elegant
p grocery_checkout.[](0)
p grocery_checkout[0]
#the second statement is syntatic sugar due to it being more elegant
p grocery_checkout.[]=(0, "Grace")
p grocery_checkout[0] = "Grace"
```

## Exceptions and Error Handling
Exceptions are the means of telling the caller that your code cannot do what was asked.

```ruby
def sqrt(num)
  unless num >= 0
    raise ArgumentError.new "Cannot take sqrt of negative number"
  end

  # code to calculate square root...
end
```

If we have a method that is taking in a argument that was not intended for it, we can **raise** an **exception** instead of returning an answer. When an exception is raised, the method stops executing. Instead of returning, an exception is thrown.

```ruby
def main
  # run the main program in a loop
  while true
    # get an integer from the user
    puts "Please input a number"
    num = gets.to_i

    begin
      sqrt(num)
    rescue ArgumentError => e
      puts "Couldn't take the square root of #{num}"
      puts "Error was: #{e.message}"
    end
  end
end
```

When the user feeds a negative number, into the **sqrt method**, it will raise an exception. Because the **main method** has wrapped the code inside a begin/rescue/end block, the exception will be caught. The code will jump to the rescue block that anticipates an **Argument Error**. It will save the exception in the variable **e** and then run the error handling code.

Ensure can be used when there is a code that must be executed whether an exception is raised or otherwise.

```ruby
begin
  a_dangerous_operation
rescue StandardError => e
  puts "Something went wrong: #{e.message}"
ensure
  puts "No matter what, make sure to execute this!"
end
```

A common example of ensuring is when closing a file.

```ruby
f = File.open
begin
  f << a_dangerous_operation
ensure
  # must. close. file.
  f.close
end
```

Retry can be used as a response to an error. This allows you retry the code until a right argument is passed in.

```ruby
def prompt_name
  puts "Please input a name:"
  # split name on spaces
  name_parts = gets.chomp.split

  if name_parts.count != 2
    raise "Uh-oh, finnicky parsing!"
  end

  name_parts
end

def echo_name
  begin
    fname, lname = prompt_name

    puts "Hello #{fname} of #{lname}"
  rescue
    puts "Please only use two names."
    retry
  end
end
```

The **retry** keyword will cause Ruby to repeat the **begin** block from the beginning. It is used to 'loop' until an operation completes successfully.

We can also implicitly write **begin/end** blocks so you only need to write **rescue**. This allows your error handling to apply to the whole method.

```ruby
def slope(pos1, pos2)
  (pos2.y - pos1.y) / (pos2.x - pos1.x)
rescue ZeroDivisionError
  nil
end
```

```ruby
def echo_name
  fname, lname = prompt_name

  puts "Hello #{fname} of #{lname}"
rescue
  puts "Please only use two names."
  retry
end
```